File: frontend/src/lib/api/candidate.ts
/**
 * API Candidat — mappée sur les endpoints backend /api/v1/candidates/*
 * La conversion snake_case ↔ camelCase est automatique via les intercepteurs du client
 */
import apiClient from './client'

// ==================== TYPES ====================

export interface ParentContact {
    id?: string
    name: string
    phone?: string
    email?: string
}

export interface AcademicRecord {
    id?: string
    trimester: number
    average: number
}

export interface SubjectScore {
    id?: string
    subject: string
    score: number
}

export interface QCMResult {
    id?: string
    score: number
    timeSpent?: number
    completedAt?: string
}

export interface Bulletin {
    id: string
    fileUrl: string
    trimester?: number
    label?: string
}

export interface SchoolInfo {
    id: string
    name: string
    city?: string
    region?: string
}

export type CandidateStatus =
    | 'registered'
    | 'qcm_pending'
    | 'qcm_completed'
    | 'regional_selected'
    | 'bootcamp_selected'
    | 'national_finalist'
    | 'rejected'

export type Gender = 'male' | 'female'

export interface CandidateProfile {
    id: string
    userId: string

    // Informations personnelles
    dateOfBirth?: string
    gender?: Gender
    phone?: string
    address?: string
    photoUrl?: string

    // Informations scolaires
    schoolId?: string
    schoolRef?: SchoolInfo
    grade?: string

    // Sous-entités normalisées
    parentContact?: ParentContact
    academicRecords: AcademicRecord[]
    subjectScores: SubjectScore[]
    qcmResult?: QCMResult
    bulletins: Bulletin[]

    // Statut et progression
    status: CandidateStatus

    createdAt: string
    updatedAt: string
}

export interface CandidateProfileUpdate {
    dateOfBirth?: string
    gender?: Gender
    phone?: string
    address?: string
    schoolId?: string
    grade?: string
    parentContact?: ParentContact
    academicRecords?: AcademicRecord[]
    subjectScores?: SubjectScore[]
}

export interface CandidateDashboard {
    user: {
        id: string
        email: string
        firstName: string
        lastName: string
        role: string
        isVerified: boolean
        isActive: boolean
        createdAt: string
        profileId?: string
    }
    profile: CandidateProfile
    progressPercentage: number
    nextSteps: string[]
    notifications: string[]
    profileCompletion: number
}

export interface UploadResponse {
    fileUrl: string
    bulletinId?: string
}

// ==================== API CALLS ====================

export const candidateApi = {
    /**
     * Récupérer le profil du candidat connecté
     * GET /candidates/me/profile
     */
    getMyProfile: async (): Promise<CandidateProfile> => {
        const { data } = await apiClient.get('/candidates/me/profile')
        return data
    },

    /**
     * Mettre à jour le profil du candidat connecté
     * PUT /candidates/me/profile
     */
    updateMyProfile: async (profileData: CandidateProfileUpdate): Promise<CandidateProfile> => {
        const { data } = await apiClient.put('/candidates/me/profile', profileData)
        return data
    },

    /**
     * Récupérer le tableau de bord du candidat
     * GET /candidates/me/dashboard
     */
    getMyDashboard: async (): Promise<CandidateDashboard> => {
        const { data } = await apiClient.get('/candidates/me/dashboard')
        return data
    },

    /**
     * Upload de la photo d'identité
     * POST /candidates/me/photo
     * Formats: JPEG, PNG — Max: 5 MB
     */
    uploadPhoto: async (file: File): Promise<UploadResponse> => {
        const formData = new FormData()
        formData.append('file', file)
        const { data } = await apiClient.post('/candidates/me/photo', formData, {
            headers: { 'Content-Type': 'multipart/form-data' },
        })
        return data
    },

    /**
     * Upload d'un bulletin scolaire
     * POST /candidates/me/bulletins
     * Format: PDF — Max: 10 MB — Max 3 bulletins
     */
    uploadBulletin: async (
        file: File,
        trimester?: number,
        label?: string
    ): Promise<UploadResponse> => {
        const formData = new FormData()
        formData.append('file', file)
        const params: Record<string, string> = {}
        if (trimester !== undefined) params.trimester = String(trimester)
        if (label) params.label = label

        const { data } = await apiClient.post('/candidates/me/bulletins', formData, {
            headers: { 'Content-Type': 'multipart/form-data' },
            params,
        })
        return data
    },

    /**
     * Supprimer un bulletin
     * DELETE /candidates/me/bulletins/:bulletinId
     */
    deleteBulletin: async (bulletinId: string): Promise<void> => {
        await apiClient.delete(`/candidates/me/bulletins/${bulletinId}`)
    },

    /**
     * Créer ou récupérer une école
     * POST /candidates/schools
     */
    createOrGetSchool: async (name: string, region: string): Promise<SchoolInfo> => {
        const { data } = await apiClient.post('/candidates/schools', {
            name,
            city: null,
            region,
        })
        return data
    },
}

File: backend/app/schemas/candidate.py
"""
Schémas Pydantic pour les candidats - Section 3.2
Adaptés à la structure 3FN
"""
from pydantic import BaseModel, Field, validator
from typing import Optional, List, TYPE_CHECKING
from datetime import date
from app.models.candidate_profile import Gender, CandidateStatus

if TYPE_CHECKING:
    from app.schemas.user import UserResponse


# === Sous-entités normalisées ===

class ParentContactCreate(BaseModel):
    """Création/mise à jour du contact parent"""
    name: str = Field(..., max_length=100)
    phone: Optional[str] = Field(None, max_length=20)
    email: Optional[str] = None


class ParentContactResponse(BaseModel):
    """Réponse contact parent"""
    id: str
    name: str
    phone: Optional[str]
    email: Optional[str]

    class Config:
        from_attributes = True


class AcademicRecordCreate(BaseModel):
    """Création d'un relevé trimestriel"""
    trimester: int = Field(..., ge=1, le=3)
    average: float = Field(..., ge=0, le=20)


class AcademicRecordResponse(BaseModel):
    """Réponse relevé trimestriel"""
    id: str
    trimester: int
    average: float

    class Config:
        from_attributes = True


class SubjectScoreCreate(BaseModel):
    """Création d'une note par matière"""
    subject: str = Field(..., min_length=2, max_length=50)
    score: float = Field(..., ge=0, le=20)


class SubjectScoreResponse(BaseModel):
    """Réponse note par matière"""
    id: str
    subject: str
    score: float

    class Config:
        from_attributes = True


class QCMResultResponse(BaseModel):
    """Réponse résultat QCM"""
    id: str
    score: float
    time_spent: Optional[int]
    completed_at: Optional[str]

    class Config:
        from_attributes = True


class BulletinResponse(BaseModel):
    """Réponse bulletin scolaire"""
    id: str
    file_url: str
    trimester: Optional[int]
    label: Optional[str]

    class Config:
        from_attributes = True


class SchoolResponse(BaseModel):
    """Réponse école"""
    id: str
    name: str
    city: Optional[str]
    region: Optional[str]

    class Config:
        from_attributes = True


class SchoolCreate(BaseModel):
    """Création d'une école"""
    name: str = Field(..., min_length=2, max_length=200)
    city: Optional[str] = None
    region: Optional[str] = None


# === Profil candidat principal ===

class CandidateProfileUpdate(BaseModel):
    """Mise à jour du profil candidat - Section 3.2"""
    # Informations personnelles
    date_of_birth: Optional[date] = None
    gender: Optional[Gender] = None
    phone: Optional[str] = Field(None, max_length=20)
    address: Optional[str] = None

    # Informations scolaires - référence par ID
    school_id: Optional[str] = None
    grade: Optional[str] = Field(None, max_length=50)

    # Contact parent (optionnel, mis à jour via endpoint dédié aussi)
    parent_contact: Optional[ParentContactCreate] = None

    # Moyennes trimestrielles
    academic_records: Optional[List[AcademicRecordCreate]] = None

    # Notes par matière
    subject_scores: Optional[List[SubjectScoreCreate]] = None


class CandidateProfileResponse(BaseModel):
    """Réponse profil candidat complet"""
    id: str
    user_id: str

    # Informations personnelles
    date_of_birth: Optional[date]
    gender: Optional[Gender]
    phone: Optional[str]
    address: Optional[str]
    photo_url: Optional[str]

    # Informations scolaires
    school_id: Optional[str]
    school_ref: Optional[SchoolResponse] = None
    grade: Optional[str]

    # Sous-entités normalisées
    parent_contact: Optional[ParentContactResponse] = None
    academic_records: List[AcademicRecordResponse] = []
    subject_scores: List[SubjectScoreResponse] = []
    qcm_result: Optional[QCMResultResponse] = None
    bulletins: List[BulletinResponse] = []

    # Statut et progression
    status: CandidateStatus

    created_at: str
    updated_at: str

    @validator('created_at', 'updated_at', pre=True)
    def convert_datetime_to_str(cls, value):
        """Convertir datetime en string ISO format"""
        if hasattr(value, 'isoformat'):
            return value.isoformat()
        return value

    class Config:
        from_attributes = True


class CandidateDashboard(BaseModel):
    """Tableau de bord candidat - Section 3.4"""
    user: "UserResponse"
    profile: CandidateProfileResponse
    progress_percentage: float
    next_steps: List[str]
    notifications: List[str]
    profile_completion: float

    class Config:
        from_attributes = True


class UploadResponse(BaseModel):
    """Réponse après upload de fichier"""
    file_url: str
    bulletin_id: Optional[str] = None  # ID de l'entrée bulletin créée

    class Config:
        from_attributes = True


# Rebuild models to resolve forward references
from app.schemas.user import UserResponse
CandidateDashboard.model_rebuild()
